### ISSUE TITLE
🐛 Bug Report: Permissions version 0.12.0

### ISSUE THREAD

**Issue by reyesmfabian** [appwrite/appwrite#2569]
### 👟 Reproduction steps

I try to create a document with the following permissions:

```
["team:" + this.adminTeam, "user:" + this.userid],
["team:" + this.adminTeam, "user:" + this.userid]
```
Now, if the user in the second permission is not in the adminTeam the console displays this error:

"AppwriteException: Read permissions must be one of: (role:all, role:member, user:60529bad133f8, team:60529c9ac57aa, team:60529c9ac57aa/owner, team:60529ca05ad80, team:60529ca05ad80/owner)"

If I change the userid to one of the adminTeam users, it works.

### 👍 Expected behavior

The document should be added correctly with the permissions, as there are users that do not belong to a specific team and still may have permissions on that particular document.

### 👎 Actual Behavior

You cannot add team and user permissions simultaneously if the users do not belong to those teams.

### 🎲 Appwrite version

Different version (specify in environment)

### 💻 Operating system

Something else

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by eldadfux** [comment]
Is `"user:" + this.userid` equals to `user:60529bad133f8`? Otherwise, for security reasons Appwrite does not allow you to assign permissions to roles your current user don't own. Meaning you can assign roles to other users from the client SDK.

This was designed to not allow users to push data for other users in the DB without their permissions. Imagine a social media feed or user notifications, you don't want user X to spam user Y. If this is the case (which I'm not sure), you can use a cloud function for setting the extra permissions.

**Comment by reyesmfabian** [comment]
> Is `"user:" + this.userid` equals to `user:60529bad133f8`? Otherwise, for security reasons Appwrite does not allow you to assign permissions to roles your current user don't own. Meaning you can assign roles to other users from the client SDK.
> 
> This was designed to not allow users to push data for other users in the DB without their permissions. Imagine a social media feed or user notifications, you don't want user X to spam user Y. If this is the case (which I'm not sure), you can use a cloud function for setting the extra permissions.

Hi @eldadfux, Is this a change from version 0.12.0? I currently have 2 solutions in version 0.11.0 and it works correctly for me this scenario:

Multistore Solution:

- Super admin user: You must create the users that will manage the stores, then create the store document and assign permissions to the newly created user. 

For this, I assign the permissions as above: 
"team:" + this.adminTeam > Super admin team.
"user:" + this.userid > Store owner user

At this point, only the super admin team and the store owner have permissions to edit this document.

In version 0.11.0 I do it without problems with the Account endpoint.

**Comment by eldadfux** [comment]
This should have been the behavior for <0.12 as well. Anyway, the team will dive deeper into this to validate no unexpected behavior is happening.

**Comment by s3ppo** [comment]
@eldadfux
i have the same problem.. in my case the team owner starts a calculation and wants to provide a result which is only for a specific user and not for the whole team...
but it throws an error that the owner is not allowed to set read permissions for a specific user

**Comment by reyesmfabian** [comment]
@eldadfux  @TorstenDittmann  @s3ppo  I suggest that at the moment of creating a document the user can set the permissions that he/she considers necessary, that is, the users and groups that will be allowed to read and write; later on only the users marked in that initial process will be the ones that can manipulate that document. This way it works in version 0.11 and in my opinion it works perfect, since it is not necessary to use Cloud Functions to assign permissions nor the server SDK.

**Comment by eldadfux** [comment]
> @eldadfux i have the same problem.. in my case the team owner starts a calculation and wants to provide a result which is only for a specific user and not for the whole team... but it throws an error that the owner is not allowed to set read permissions for a specific user
> 
> or an other example is a chat.. user1 sends a private message to user2.. but user1 is now allowed to set read permission `user:user2`
> 
> in 0.11 this was possible

Sounds like we might had a bug in 0.11 because this was the designed behavior for a long time now, but we'll double and triple check this report internally.

One of our suggestions for such a use case was to use functions to initiate a document or collection that share permissions for both users (you can do that with an API key). Once both users have permissions to the same document or collections they can start to chat.

The idea was partially inspired by a classic social media friend/connection request to allow communication. For "other" messages from anonymous user a collection with `role:all` write permissions could have been an equivalent solution.

I'm very curios to know if this design can help your use case without regard to the extra research we'll make.

**Comment by reyesmfabian** [comment]
> Sounds like we might had a bug in 0.11 because this was the designed behavior for a long time now, but we'll double and triple check this report internally.
> 
> One of our suggestions for such a use case was to use functions to initiate a document or collection that share permissions for both users (you can do that with an API key). Once both users have permissions to the same document or collections they can start to chat.
> 
> The idea was partially inspired by a classic social media friend/connection request to allow communication. For "other" messages from anonymous user a collection with `role:all` write permissions could have been an equivalent solution.
> 
> I'm very curios to know if this design can help your use case without regard to the extra research we'll make.

Hi @eldadfux One question, to be clear, the current behavior of version 0.12 is the definitive one? or is there a possibility to adapt it as it was in version 0.11? I mean, from now on when using the Account endpoint you can only assign permissions to a document for the user who creates it and his assigned teams? All additional permissions must be added through the ApiKey? Best regards

**Comment by eldadfux** [comment]
Gonna double and triple check this with the team on Monday before we can share any findings 🙏

**Comment by reyesmfabian** [comment]
Hi, I'm not trying to sound intense or annoying with this post, but, have you managed to find related information?

**Comment by eldadfux** [comment]
@reyesmfabian not annoying at all, this is on a high priority, we have a call about this issue tomorrow.

**Comment by eldadfux** [comment]
Update: we had quite a few discussion on this one internally. From all we know if something didn't work in <0.11 it was probably a bug and not the intended behavior. 

With the introduction of collection level permission we thought it might allow us to be more permissive and allow users to set permission they don't own, but trying to validate this theory on multiple common use cases for building things with Appwrite kept getting us to the same concern. Allowing something like this will allow an easy way to abuse sets of data by polluting the datasets with permissions for users who don't want them.

A good example to try and emphasize this is something like Dropbox, imagine you start your drive and find 1,000 shared documents that you don't care about and **with no** easy way to filter them away from your documents.

That said, we understand the difficulty. Right now the only way to share documents is using an API key or a Cloud Function. Functions on Appwrite 0.13 would be more intuitive to use with the introduction of sync execution, but writing a function still look like a big ask for something that common. That is why we started a discussion about a potential Permissions API to easily share access between users of your app.

More info on the new suggested feature is available in this discussion: https://github.com/appwrite/appwrite/discussions/2689 - your feedback is crucial, we'd love you to participate!

**Comment by reyesmfabian** [comment]
Hi @eldadfux I wrote a comment in this thread #2689 about my experience with Appwrite which is now going on about a year. 

As a summary, I'm not a fan of new permissions per collection, as this somewhat limits being specific with permissions at the document level to avoid situations like you mention with Dropbox. 

Now, as I wrote there, an invitation system would be fine for social network type applications and user interactions, but, for more enterprise environments it is not very effective and well seen; there we are looking for users to have the least amount of prompts from the backend and limit themselves to do the function for which their role within the application is designed. 

I love Appwrite and I am using it in some production applications, but I have already had several headaches when updating versions and I have had to rewrite my applications several times and most of them have been due to permissions issues. I have several issues created by the same thing in different versions. Now, I am worried about upgrading my servers to version 0.12 to take advantage of the new features such as Querys and performance improvements knowing that I have to rewrite practically all my code without the certainty of not having to do the same in future 0.13 or 0.14 and my applications will break again.

I think it is necessary to define a final permissions scheme as Firebase did in its time and marry it.

I would love to help with debugging and reviewing versions together with the team, but unfortunately I need that time to survive in this part of the world 😅

For now, I think the best recommendation if I want to upgrade to version 0.12 is to assign permissions from the backend or using cloud functions, am I right?
